lib/kubernetes-deploy/deploy_task.rb

```diff
@@ -267,23 +271,26 @@ def check_initial_status(resources)
     measure_method(:check_initial_status, "initial_status.duration")
 
     def secrets_from_ejson
-      ejson_provisioner.resources
+      ejson_provisioners.flat_map(&:resources)
     end
 
     def discover_resources
       resources = []
       crds = cluster_resource_discoverer.crds.group_by(&:kind)
       @logger.info("Discovering templates:")
 
-      TemplateDiscovery.new(@template_dir).templates.each do |filename|
-        split_templates(filename) do |r_def|
-          crd = crds[r_def["kind"]]&.first
-          r = KubernetesResource.build(namespace: @namespace, context: @context, logger: @logger, definition: r_def,
-            statsd_tags: @namespace_tags, crd: crd)
-          resources << r
-          @logger.info("  - #{r.id}")
+      TemplateDiscovery.new(@template_dirs).templates.each do |template_dir, filenames|
```

Comment:
I'm :+1: on pushing the template splitting to the renderer. I don't think `TemplateDiscovery` should own resource instantiation, however.

IMO, `TemplateDiscovery` shouldn't own anything other than returning a list of valid template files, organized by directory. For now, `deploy_task` is the only thing that actually needs to instantiate resources, so I'm ok with having such instantiation logic live there.

TBH, the more I think about it the more I think we should move rendering out of `deploy_task`. Coupling rendering with deployment is a bad anti-pattern (as we've discussed ad nauseam).

As it stands, I'm pretty comfortable with the current solution of creating a renderer-constructor in `DeployTask#new` and opportunistically creating renderers from provided template dirs at their actual point of use. Moving template splitting into the renderer itself jives well with this approach and let's us kick the tougher question of whether we should be rendering here at all down the road.

Reply:
I have a few more ideas after thinking about this a bit more.

Perhaps `TemplateDiscovery` is poorly named, and should instead be something like `ResourceDiscovery`. `ResourceDiscovery` is then responsible for finding all relevant resource files in a template directories, splitting the templates in each file, calling the renderer on them, and returning a list of fully realized `KubernetesResource`s.

This allows us to completely remove any explicit references to the renderer in `DeployTask`. This is beneficial since rendering logic shouldn't need to be directly in `DeployTask`, anyway (it should only care about fully realized resources). Also, if we ever decide to disallow rendering of ERB inside `DeployTask`, we can simple pass in `render_erb: false` and ignore any `.erb` files. WDYT?